src/lib/rest/RestService.cpp

```diff
@@ -52,9 +52,111 @@
 #define MAX_TENANT_NAME_LEN_STRING    "50"
 
 
+static std::string textParseAttributeValue(ConnectionInfo* ciP, ContextAttribute* caP)
+{
+  double d;
+
+  // 1. Starts and ends with citation marks?
+  if (ciP->payload[0] == '"')
+  {
+    char* end = &ciP->payload[strlen(ciP->payload) - 1];
+
+    if (*end == '"')
+    {
+      *end = 0;
+      caP->stringValue = &ciP->payload[1];
+      caP->valueType   = orion::ValueTypeString;
+    }
+    else
+    {
+      OrionError oe(SccBadRequest, "Missing citation-mark at end of string");
+
+      return oe.render(ciP, "");
+    }
+  }
+
+  // 2. True or false?
+  else if ((strlen(ciP->payload) == 4) && ((strcmp(ciP->payload, "true") == 0) || (strcmp(ciP->payload, "True") == 0) || (strcmp(ciP->payload, "TRUE") == 0)))
+  {
+    caP->boolValue   = true;
+    caP->valueType   = orion::ValueTypeBoolean;
+  }
+  else if ((strlen(ciP->payload) == 5) && ((strcmp(ciP->payload, "false") == 0) || (strcmp(ciP->payload, "False") == 0) || (strcmp(ciP->payload, "FALSE") == 0)))
+  {
+    caP->boolValue   = false;
+    caP->valueType   = orion::ValueTypeBoolean;
+  }
+
+  // 3. Null ?
+  else if ((strlen(ciP->payload) == 4) && ((strcmp(ciP->payload, "null") == 0) || (strcmp(ciP->payload, "Null") == 0) || (strcmp(ciP->payload, "NULL") == 0)))
+  {
+    caP->valueType   = orion::ValueTypeNull;
+  }
+
+  //
+  // 4. Valid Double?
+  // FIXME P4: this is much more complex than just (atof(string) != 0 || string == "0")
+  //           0.000 also is a valid float and it given 0 - 0.0000000000, and 0e0 also ...
+  //           And, even worse, 123K gives 123.0 back, we would need to analyze the string to try to
+  //           find garbage bytes after it if we want to detect this error.
+  //           However, all of this is not so extremely important and for now, (atof(string) != 0 || string== "0") is good enough
+  //
+  else if (((d = strtod(ciP->payload, NULL)) != 0) || ((ciP->payload[0] == '0') || (ciP->payload[1] == 0)))
```

Comment:
On 01/10/15 00:22, Fermín Galán Márquez wrote:

In src/lib/rest/RestService.cpp<https://github.com/telefonicaid/fiware-orion/pull/1299#discussion_r40860178>:

> +
> +  // 3. Null ?
> +  else if ((strlen(ciP->payload) == 4) && ((strcmp(ciP->payload, "null") == 0) || (strcmp(ciP->payload, "Null") == 0) || (strcmp(ciP->payload, "NULL") == 0)))
> +  {
> +    caP->valueType   = orion::ValueTypeNull;
> +  }
> +
> +  //
> +  // 4. Valid Double?
> +  // FIXME P4: this is much more complex than just (atof(string) != 0 || string == "0")
> +  //           0.000 also is a valid float and it given 0 - 0.0000000000, and 0e0 also ...
> +  //           And, even worse, 123K gives 123.0 back, we would need to analyze the string to try to
> +  //           find garbage bytes after it if we want to detect this error.
> +  //           However, all of this is not so extremely important and for now, (atof(string) != 0 || string== "0") is good enough
> +  //
> +  else if (((d = strtod(ciP->payload, NULL)) != 0) || ((ciP->payload[0] == '0') || (ciP->payload[1] == 0)))


I guess that in this case timing information is not necessary (specially if it costly to calculate that) if you have an argument that support the efficient improvement (e.g. 4 functional calls vs. 1).

—
Reply to this email directly or view it on GitHub<https://github.com/telefonicaid/fiware-orion/pull/1299/files#r40860178>.

Its easy enough to measure (I have it implemented for other cases already),
It's really just a few calls to clock_gettime and a function I implemented for the semaphore waiting times - clock_difftime.
Takes ... 10 min tops.
It is interesting to know the difference ...

________________________________

Este mensaje y sus adjuntos se dirigen exclusivamente a su destinatario, puede contener información privilegiada o confidencial y es para uso exclusivo de la persona o entidad de destino. Si no es usted. el destinatario indicado, queda notificado de que la lectura, utilización, divulgación y/o copia sin autorización puede estar prohibida en virtud de la legislación vigente. Si ha recibido este mensaje por error, le rogamos que nos lo comunique inmediatamente por esta misma vía y proceda a su destrucción.

The information contained in this transmission is privileged and confidential information intended only for the use of the individual or entity named above. If the reader of this message is not the intended recipient, you are hereby notified that any dissemination, distribution or copying of this communication is strictly prohibited. If you have received this transmission in error, do not read it. Please immediately reply to the sender that you have received this communication in error and then delete it.

Esta mensagem e seus anexos se dirigem exclusivamente ao seu destinatário, pode conter informação privilegiada ou confidencial e é para uso exclusivo da pessoa ou entidade de destino. Se não é vossa senhoria o destinatário indicado, fica notificado de que a leitura, utilização, divulgação e/ou cópia sem autorização pode estar proibida em virtude da legislação vigente. Se recebeu esta mensagem por erro, rogamos-lhe que nos o comunique imediatamente por esta mesma via e proceda a sua destruição